File: kattis/almostunionfind.cpp
#include <bits/stdc++.h>
using namespace std;

class UnionFind {
private:
    vector<long long> p, rank, setSize, sumset;
    int numSets;

public:
    UnionFind(int N) {
        setSize.assign((2*N)+2, 1);
        numSets = N;
        rank.assign((2*N)+2,0);
        p.assign((2*N)+2,0);
        sumset.assign((2*N)+2, 0);
        for (int i = 0; i <= (2*N)+1; i++) {
        	if (i > N) {
        		p[i] = i;
        	}
            else {
            	p[i] = N+i+1;
            	sumset[N+i+1] = i;
            }
        }
    }
    int findSet(int i) {
        return (p[i] == i) ? i : (p[i] = findSet(p[i]));
    }
    bool isSameSet(int i, int j) {
        return findSet(i) == findSet(j);
    }
    void unionSet(int i, int j) {
        if(!isSameSet(i,j)) {
            numSets--;
            int x = findSet(i);
            int y = findSet(j);
            if (rank[x] > rank[y]) {
                p[y] = x;
                setSize[x] += setSize[y];
                sumset[x] += sumset[y];
            } 
            else {
                p[x] = y;
                setSize[y] += setSize[x];
                sumset[y] += sumset[x];
                if(rank[x] == rank[y]) {
                    rank[y]++;
                }
            }
        }
    }
    void Move(int i, int j) {
    	if (!isSameSet(i,j)) {
    		int x = findSet(i);
    		int y = findSet(j);
    		setSize[x]--;
    		setSize[y]++;
    		p[i] = y;
    		sumset[x] -= i;
    		sumset[y] += i;
    	}
    }
    int numDisjointSets() {
        return numSets;
    }
    int sizeOfSet(int i) {
        return setSize[findSet(i)];
    }
    long long sumofset(int i) {
    	return sumset[findSet(i)];
    }
};

int main() {
	int n, m;
	while (cin >> n >> m) {
		UnionFind uf(n);
		for (int i = 0; i < m; i++) {
			int comm;
			cin >> comm;
			if (comm == 1) {
				int a, b;
				cin >> a >> b;
				uf.unionSet(a, b);
			}
			else if (comm == 2) {
				int a, b;
				cin >> a >> b;
				uf.Move(a, b);
			}
			else if (comm == 3) {
				int a;
				cin >> a;
				cout << uf.sizeOfSet(uf.findSet(a)) << ' ' << uf.sumofset(uf.findSet(a)) << endl;
			}			
		}
	}
}

File: kattis/almostunionfind.java
import java.util.*;

class UFDS {
    private ArrayList<Integer> parents;
    private ArrayList<Integer> ranks;
    private ArrayList<Integer> size_of_set;
    private ArrayList<Long> sum_of_set;
    private int num_of_set;

    public UFDS(int n) {
        this.parents = new ArrayList<>();
        this.ranks = new ArrayList<>();
        this.size_of_set = new ArrayList<>();
        this.sum_of_set = new ArrayList<>();
        this.num_of_set = n;

        for (int i = 0; i <= 2*n + 1; i++) {
            this.ranks.add(0);
            this.size_of_set.add(1);
            if (i > n) {
                this.parents.add(i);
                this.sum_of_set.add(((long)i - n - 1));
            } else {
                this.parents.add(n + i + 1);
                this.sum_of_set.add(0L);
            }
        }
    }

    public int findSet(int x) { 
        if (parents.get(x) == x)  {
            return x;
        } else {
            int value = findSet(this.parents.get(x)); 
            parents.set(x, value);
            return value; 
        } 
    }

    public Boolean isSameSet(int x, int y) { 
        return findSet(x) == findSet(y); 
    }

    public void unionSet(int m, int n) { 
        if (!isSameSet(m, n)) { 
            this.num_of_set--; 
            int x = findSet(m);
            int y = findSet(n);

            if (this.ranks.get(x) > this.ranks.get(y)) { 
                this.size_of_set.set(x, this.size_of_set.get(x) + this.size_of_set.get(y)); 
                this.sum_of_set.set(x, this.sum_of_set.get(x) + this.sum_of_set.get(y));
                this.parents.set(y, x);
            } else { 
                this.size_of_set.set(y, this.size_of_set.get(y) + this.size_of_set.get(x));
                this.sum_of_set.set(y, this.sum_of_set.get(x) + this.sum_of_set.get(y));
                this.parents.set(x, y); 
                if (this.ranks.get(x) == this.ranks.get(y)) { 
                    this.ranks.set(y, this.ranks.get(y) + 1); 
                } 
            } 
        }
    }

    public int numOfDisjointSets() { 
        return this.num_of_set; 
    }
    
    public int sizeOfSet(int x) { 
        return this.size_of_set.get(findSet(x)); 
    }

    public void move(int from, int to) {
        if (!isSameSet(from, to)) {
            int x = this.parents.get(from);
            int y = this.parents.get(to);
            this.size_of_set.set(x, this.size_of_set.get(x) - 1);
            this.size_of_set.set(y, this.size_of_set.get(y) + 1);
            this.parents.set(from, y);
            this.sum_of_set.set(x, this.sum_of_set.get(x) - from);
            this.sum_of_set.set(y, this.sum_of_set.get(y) + from);
        }
    }

    public long sumSet(int x) {
        return sum_of_set.get(findSet(x));
    }
}

class Main {
    public static void main(String[] args) {
        Kattio io = new Kattio(System.in);

        while (io.hasMoreTokens()) {
        	int n = io.getInt();
	        int m = io.getInt();

	        UFDS ufds = new UFDS(n);

	        for (int i = 0; i < m; i++) {
	            int op = io.getInt();
	            if (op == 1) {
	                ufds.unionSet(io.getInt(), io.getInt());
	            } else if (op == 2) {
	                ufds.move(io.getInt(), io.getInt());
	            } else {
	                int x = io.getInt();
	                io.println(ufds.sizeOfSet(x) + " " + ufds.sumSet(x));
	            }
	        }
        }
        
        io.flush();
    }
}